### ISSUE TITLE
Abbility to define a global routing gaurd

### ISSUE THREAD

**Issue by yourikoeman** [nareshbhatia/mobx-state-router#143]
First of all, great project and many thanks for maintaining it.

I have run into a issue where i would like to be able to define a global `beforeEnter` style routing gaurd. specifically
for handling authentication and have it run on every route (so you don't have to rely on making a wrapper function used on every other routing gaurd). 

This could be implemented internally without breaking any of this libraries APIs.

With for example the following way of calling it
```
// RouterStore.js
class RouterStore {
    beforeEach (from: RouterState, to: RouterState): Promise<RouterState>
}

functionally this should be very similar to the behaviour of the routerstate BeforeEnter, 
but be applied to all routes

```

Is this something that if i implement it, you would consider merging into the mainline?

**Comment by nareshbhatia** [comment]
@yourikoeman, glad to hear that you are finding this project useful.

Regarding your issue, have you seen this recipe in the docs: https://nareshbhatia.github.io/mobx-state-router/docs/guides-recipes#redirecting-to-the-sign-in-page? Does it address your problem?

**Comment by yourikoeman** [comment]
Thanks for taking your time to reply :)

The behavior would prevent public routes from being defined on the routes beforeEnter in this case yes.
But the place where i would define the behaviour for what routes should be considered "public" would be in the beforeEach

Like E.G (skeleton taken from one of the examples in this repo)
```js
//PseudoCode 
const routerStore = new RouterStore(
        routes,
        notFound,
        {
            beforeEach: async (from, to) => {
                const publicRoutes = [
                    'login',
                    'register',
                    'reset-pass',
                    'ETC'
                ];

                let publicRouteMatched = publicRoutes.find((routeName) => routeName === to.name);
                if (publicRouteMatched) {
                    // Same behaviour upon returning void
                    // as other routing guards
                    return;
                }

                // Check authentication from store
                // Pretend there is code here

                let authenticated;
                if (authenticated) {
                    return
                }
                
                // Handle not authenticated case, E.G redirect to login
            }
        }
    );
}
```

The main advantage is when you have behavior you know for sure needs to be executed every routing change
you can define it globally. 

if someone configures the per route guards wrong the application will still function with a
"bug". 

In the proposed change if applied it takes explicit action to allow new routes to be unprotected instead of
other way around (and developers can forget to add the wrapper);

**Comment by nareshbhatia** [comment]
@yourikoeman, I see your point now. Please feel free to send a PR.

To confirm my understanding:
1. You will add a global `beforeEnter` hook.
2. This will be added as a special option which can be passed into the 3rd parameter of `RouterStore`. Type will be `TransitionHook`.
3. Can you please name this option `globalBeforeEnter`, just so that people don't get confused with `Route.beforeEnter`?

Please keep changes to minimal necessary and make sure that all existing tests are running successfully.

**Comment by yourikoeman** [comment]
Yes, i will incorporate those changes and the name change of said option. 
During this weekend i will implement it and reference the PR in this thread.

I will also make sure to update the tests, documentation and examples. 

Are there special considerations i need to take when updating those?

**Comment by nareshbhatia** [comment]
Thanks @yourikoeman.

No special considerations regarding updating of tests and documentation - just an overall focus on quality. I will of course be reviewing to make sure everything is in order.

**Comment by yourikoeman** [comment]
Pull request #144 has been made
